server/src/main/java/org/elasticsearch/search/SearchHit.java

```diff
@@ -543,6 +605,8 @@ public void setInnerHits(Map<String, SearchHits> innerHits) {
         static final String _PRIMARY_TERM = "_primary_term";
         static final String _SCORE = "_score";
         static final String FIELDS = "fields";
+        static final String DOCUMENT_FIELDS = "document_fields";
```

Comment:
HI @rjernst , I think the issue I struggled with last time and tried to describe in https://github.com/elastic/elasticsearch/pull/41656#discussion_r292000298 is that it is difficult to discern metafields from regular fields during deserialization if we store both of them under the same key "fields". Then the only source of information would be `MapperService.getAllMetaFields()` which we are trying to refactor away. Do you see any other way to separate those groups of fields during deserialization if json structure remains the same?

Reply:
Hi @rjernst, I was reading through code again - looks like that concern of mine is no big issue after all.

> I think there are 2 cases here: rest (json) responses, and internal node to node communication. The latter is handled by the stream input/output version 

agree. The node to node communication was already addressed before in this PR, we are only discussing json responses here

> For rest/json response, we currently recommend users upgrade Elasticsearch before upgrading the HLRC, which means the rest layer here must stay the same for compatibility with the older client during this upgrade period.

I previously though it would be difficult to have it compatible, however after having another look, I don't see problems with that. Currently this PR should not be introducing breaking changes to rest/json api.

>Only document fields are output within the "fields" object. 

Is preserved.

> Meta fields are output at the root level (the same level as "fields"). 
 
Is preserved as well.

> We should be able to catch the rest of the special fields we did not parse specially (ie if a field wasn't parsed explicitly like _id, then it is a meta field).

Exactly this is happening in the `declareMetaDataFields` method which already existed before this PR. Currently this method is mainly remaining intact in the PR, I only change where in the object parser map does it store the metafields

> Currently we register each meta field in the object parser. Instead, we should now be using an UnknownFieldConsumer.

This is the only part I have questions about. Looks like `declareMetaDataFields` is doing the job already, why should we move from object parser to UnknownFieldConsumer?

Also addressed part of your comments in the last commit.

Reply:
@mayya-sharipova I was referring to the last question here.